04-cs-fundamentals/internship/binary-search-trees.md

```diff
@@ -1,23 +1,253 @@
+
+Instead of a node with one link to a next node, we can create nodes with 2 pointers, left and right.  Each node is not allowed to point back to a parent or grandparent.   So you can arrive at a child by following the links of it's parents, but there is no path from a child to a parent.  This ends up forming a tree like shape with each node having 0, 1 or 2 "children".  This is called a _Binary Tree_.  In this way a binary tree is an _acyclic, connected graph_.  It forms no cycles or loops of links, and each node is connected to others in the data structure.
+
+In a special type of Binary Tree, called a _Binary Search Tree_ the left pointer points to all elements smaller than the node.  The right pointer points to all nodes greater than the given node.   Each node can refer to other nodes, but there is no way to go from a child node to a parent.  The links go in one direction.  A Tree is hierarchical with certain nodes acting as parents to others.  
+
+![bst.png](images/bst.png)
+
+Notice that nodes 61 and 77, both form subtrees of the larger tree data structure.  This enables you to use **recursion** on most tree methods because each child node forms a subtree.
+
+**Exercise**
+
+Given the following values draw a binary search tree:
+
+"Ringo", "John", "Paul", "George"
+
+## Binary Search Tree Node
+
+Just like a Linked List node, a tree node has a data field, and links to other nodes.  However instead of linking to previous and next nodes, the links are to nodes less than and greater than the value in the data field.
+
+![Binary Search Tree Example](images/bst1.png)
+
+The Node class for a Binary Search Tree (BST) could look like this:
+
+```ruby
+class TreeNode
+  attr_reader :data
+  attr_accessor :left, :right
+
+  def initialize(data)
+    @data = data
+    @left = @right = nil
+  end
+end
+```
+
+These nodes are arranged with one node known as the **root** which is where the tree begins.  New nodes as they are added are placed to the left, if they are less than the current node, and to the right if they are greater than the current node.  This is a natually recursive process.
+
+```
+Method add:
+  if the root is nil set the root to be a new node with the given value and return the node.
+
+  Otherwise 
+    if the value is less than or equal to the current root's value, make root's left be the result of calling add on root's left.
+    otherwise make root's right be the result of calling add on root's right.
+```
+
+You can observe this in the [Binary Tree Visualizer](http://btv.melezinek.cz/binary-search-tree.html)
+
+
+## The Tree Class
+
+You can build a `Tree` class in a manner similar to how we created a `LinkedList` class.
+
+```ruby
+class Tree
+  attr_reader :root
```

Comment:
I might lay out this starter code up at the top, making it clear that we have two different classes, the nodes and the tree itself.

Are you planning to add code to this definition as you go along?